### PR TITLE
RESTWS-443: Added EncounterRoleResource1_9, EncounterProviderResource…

### PR DESCRIPTION
**Pull request by mhawila**
…1_9 and associated tests. Added support for multiple provider per encounter

to EncounterResource1_9.

Added tests to EncounterRoleController1_9Test class

### CODE REVIEW

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    @Override
+    public EncounterProvider getByUniqueId(String uniqueId) {
+        return Context.getService(RestHelperService.class).getObjectByUuid(EncounterProvider.class,uniqueId);
+    }
+
+    @Override
+    protected void delete(EncounterProvider delegate, String reason, RequestContext context) throws ResponseException {
+         throw new ResourceDoesNotSupportOperationException();
```
Comment (rkorytkowski) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java]: I think delete should call  delegate.getEncounter().removeProvider(delegate.getEncounterRole(),delegate.getProvider());

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    @Override
+    public EncounterProvider save(EncounterProvider delegate) {
+        delegate.getEncounter().getEncounterProviders().add(delegate);
+        return delegate;
+    }
+
+    @Override
+    public void purge(EncounterProvider delegate, RequestContext context) throws ResponseException {
+         delegate.getEncounter().removeProvider(delegate.getEncounterRole(),delegate.getProvider());
```
Comment (rkorytkowski) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java]: purge is actually removing an element from the list i.e. delegate.getEncounter().getEncounterProviders()

Comment (rkorytkowski) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java]: @djazayeri, do you think we should support that or EncounterProviders are only intended to be deleted (voided)?

Comment (mhawila) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java]: Or may be I could move the functionality to delete() and throw exception in purge?

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java (hunk 1 above)
Comment (mhawila) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java]: Yeah sure, will change this one.

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java (hunk 2 above)
Comment (mhawila) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java]: I think purging should not be supported?

Comment (djazayeri) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java]: If there is an existing API method to purge this, we should support it via REST also. If not, we should not.

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    /**
+     * Display string for EncounterProvider
+     *
+     * @param encounterProvider
+     * @return String uuid
+     */
+    @PropertyGetter("display")
+    public String getDisplayString(EncounterProvider encounterProvider) {
```
Comment (djazayeri) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java]: I would expect the "display" field to be the provider ID, or else provider.person.preferredName

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterResource1_9.java [hunk 4]
```diff
@@ -36,25 +39,47 @@ public DelegatingResourceDescription getRepresentationDescription(Representation
...
+        description.addProperty("encounterProviders");
 		return description;
 	}
-	
+
+    @Override
+    public Encounter save(Encounter delegate) {
+        //This is a hack to save encounterProviders correctly. Without this they are created without encounter_id in
+        //the database.
+        for(EncounterProvider ep:delegate.getEncounterProviders()){
```
Comment (djazayeri) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterResource1_9.java]: btw, please format this nicely in the OpenMRS style, e.g. spaces like

    for (EncounterProvider ep : delegate.getEncounterProviders()) {

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java (hunk 3 above)
Comment (mhawila) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterProviderResource1_9.java]: What about the associated provider role? I couldn't really think of the what would the best thing to display. May be a combination of provider ID and encounterRole name?

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterResource1_9.java (hunk 4 above)
Comment (mhawila) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/EncounterResource1_9.java]: Noted.

**Pull request merged**